packages/versioning/src/versioning.ts

```diff
@@ -0,0 +1,317 @@
+    !hasVersion(p, t, v)
+  ) {
+    reportDiagnostic(p, { code: "version-not-found", target: t, format: { version: v } });
+    return;
+  }
+
+  p.stateMap(addedOnKey).set(t, v);
+}
+export function $removed(p: Program, t: Type, v: string) {
+  if (typeof v !== "string") {
+    reportDiagnostic(p, { code: "version-must-be-string", target: t });
+    return;
+  }
+  // this validation doesn't work for model properties because we can't walk up to
+  // get the container type.
+  if (
+    ["EnumMember", "ModelProperty", "UnionVariant"].indexOf(t.kind) === -1 &&
+    !hasVersion(p, t, v)
+  ) {
+    reportDiagnostic(p, { code: "version-not-found", target: t, format: { version: v } });
+    return;
+  }
+  p.stateMap(removedOnKey).set(t, v);
+}
+export function $renamedFrom(p: Program, t: Type, v: string, oldName: string) {
+  if (typeof v !== "string") {
+    reportDiagnostic(p, { code: "version-must-be-string", target: t });
+    return;
+  }
+  if (
+    ["EnumMember", "ModelProperty", "UnionVariant"].indexOf(t.kind) === -1 &&
+    !hasVersion(p, t, v)
+  ) {
+    reportDiagnostic(p, { code: "version-not-found", target: t, format: { version: v } });
+    return;
+  }
+  const record = { v: v, oldName: oldName };
+  p.stateMap(renamedFromKey).set(t, record);
+}
+
+export function $madeOptional(p: Program, t: Type, v: string) {
+  if (typeof v !== "string") {
+    reportDiagnostic(p, { code: "version-must-be-string", target: t });
+    return;
+  }
+  if (
+    ["EnumMember", "ModelProperty", "UnionVariant"].indexOf(t.kind) === -1 &&
+    !hasVersion(p, t, v)
+  ) {
+    reportDiagnostic(p, { code: "version-not-found", target: t, format: { version: v } });
+    return;
+  }
+
+  p.stateMap(madeOptionalKey).set(t, v);
+}
+
+export function getRenamedFromVersion(p: Program, t: Type) {
+  return p.stateMap(renamedFromKey).get(t)?.v ?? -1;
+}
+export function getRenamedFromOldName(p: Program, t: Type) {
+  return p.stateMap(renamedFromKey).get(t)?.oldName ?? "";
+}
+export function getAddedOn(p: Program, t: Type) {
+  return p.stateMap(addedOnKey).get(t) ?? -1;
+}
+export function getRemovedOn(p: Program, t: Type) {
+  return p.stateMap(removedOnKey).get(t) ?? Infinity;
+}
+
+export function getMadeOptionalOn(p: Program, t: Type) {
+  return p.stateMap(madeOptionalKey).get(t) ?? -1;
+}
+
+export function $versioned(p: Program, t: Type, v: Type) {
+  if (t.kind !== "Namespace") {
+    reportDiagnostic(p, { code: "versioned-not-on-namespace", target: t });
+    return;
+  }
+  const versions = [];
+
+  switch (v.kind) {
+    case "String":
+    case "Number":
+      versions.push(v.value);
+      break;
+    case "Union":
+      for (const variant of v.options) {
+        if (variant.kind === "String" || variant.kind === "Number") {
+          versions.push(variant.value);
+        }
+      }
+      break;
+    case undefined:
+      if (typeof v === "number" || typeof v === "string") {
+        versions.push(v);
+      }
+  }
```

Comment:
Should have a default case with an error diagnostic to cover the type target being something else.

Reply:
https://github.com/microsoft/cadl/issues/215